### ISSUE TITLE
Slow networking since 968ce9af

### ISSUE THREAD

**Issue by cheapie** [minetest/minetest#8183]
##### Issue type

 - Bug report
 
##### Minetest version

```
5.0.0-dev-2ae794ac
```

##### OS / Hardware

Operating system: Debian sid amd64
CPU: AMD Ryzen 3 2200G

##### Summary
I recently switched from 0.4.17.1 to 5.0.0-dev, and when I did, networking became very slow. On public servers, this is most noticeable in that excessive network lag occurs whenever anything much is happening, as well as that joining (especially the item definitions step) takes much longer than before. In singleplayer, the most noticeable problem is that node definitions take much longer.

##### Steps to reproduce
1. Build 07b1743 or earlier
2. Start a singleplayer game with a heavily-modded world (dreambuilder or similar)
3. Note the time that node definitions and media take to load (about 0.5s each in my testing)
4. Build 968ce9af or later, all the way up to the current version as of now
5. Start a singleplayer game with the same world
6. Note the time that node definitions and media take to load (about 5-10s each in my testing)

##### Random additional testing that may or may not mean anything
- In UDPPeer::reportRTT in connection.cpp, forcing timeout to always be RESEND_TIMEOUT_MIN makes joining singleplayer fast again
- In the same function, adding a line to print the timeout each time it's calculated produces an awful lot of "3.000000" even in singleplayer
- Changing RESEND_TIMEOUT_FACTOR in src/constants.h from 4 (default) to 0.4 fixes singleplayer

**Comment by Fixer-007** [comment]
Could be the reason why VanessaE 5.0 servers are _ridiculously_ slow, both had max_lag 4 and 0,2, yet map loading and chat was very very slow, with lock ups.

**Comment by VanessaE** [comment]
The odd thing is, I noticed no unusual lag when I was on.  I couldn't outrun the map loader on my Dreambuilder creative server, for example (by that I mean, it loads and renders so quickly that I never hit unloaded/ignore blocks while running in a straight line at full speed).  Interactions such as doors, writing on signs, moving things in my inventory, and dig and place nodes all seemed normal.

**Comment by cheapie** [comment]
> The odd thing is, I noticed no unusual lag when I was on.

I suspect this may be affecting those of us with worse connections more, as the calculations in question are based around RTT. My connection was good enough for a good experience with 0.4.17.1, but it's not great.

**Comment by paramat** [comment]
IRC discussion http://irc.minetest.net/minetest-dev/2019-02-06#i_5490804

**Comment by paramat** [comment]
@HybridDog

**Comment by HybridDog** [comment]
Here's code which might be related.
reportRTT changes resend_timeout:
https://github.com/minetest/minetest/blob/fc566e2e1074e501283d4be70a654d6b79ef07ff/src/network/connection.cpp#L911-L915
getResendTimeout returns this number:
https://github.com/minetest/minetest/blob/fc566e2e1074e501283d4be70a654d6b79ef07ff/src/network/connection.h#L668-L669
it is used to find timed_outs, which probably contains less items if the resend_timeout is big:
https://github.com/minetest/minetest/blob/fc566e2e1074e501283d4be70a654d6b79ef07ff/src/network/connectionthreads.cpp#L204-L222

**Comment by paramat** [comment]
On IRC we are suggesting just reverting that commit, since the RTT fix isn't essential http://irc.minetest.net/minetest-dev/2019-02-06#i_5491057

**Comment by paramat** [comment]
Anyone feel like making a PR to revert https://github.com/minetest/minetest/commit/968ce9af598024ec71e9ffb2d15c3997a13ad754 ?

**Comment by benrob0329** [comment]
Why not keep the fix but apply one or more of the timout changes?

**Comment by nerzhul** [comment]
We are near to release and this part is very touchy. It's better to revert, and reintegrate better version after release.

**Comment by ClobberXD** [comment]
PR to revert the commit in question: #8187

**Comment by ClobberXD** [comment]
I still don't understand how the RTT fixes commit affects network performance. Would somebody be able to explain exactly what causes this issue and why, in layman terms? :)

**Comment by HybridDog** [comment]
I don't understand it, too. When I changed the code I thought that RTT is only used for the debug information. RTT was broken before my commit: it calculated the average Round Trip Time incorrectly since somebody broke it a long time ago.

**Comment by cheapie** [comment]
RTT is used by UDPPeer::reportRTT() to set resend_timeout, which seems to have something to do with networking. Higher-than-usual values there dramatically lower network performance.

**Comment by HybridDog** [comment]
I assume that the timeout is only used if a packet is lost, i.e. missing acknowledgement. There shouldn't be packet loss in singleplayer, except if the operating system discards UDP packets because of flooding (if I remember correctly, this can happen with UDP but not with TCP). Maybe the RTT calculation does not work right while joining. Ingame, RTT is 0.001s (according to debug info).

I've just noticed something:
If I press ESC in singleplayer, to pause the game, and then look at the RTT (in the debug info), it is 0.001s for some time, then it grows until it reaches 5s, where it's set back to 0.001s. I don't know if this is intended behaviour.

**Comment by paramat** [comment]
Maybe something to test is to add code to set the value of RTT and see the effect of different values?

**Comment by paramat** [comment]
#8187 merged, reverted for 5.0.0.
Leaving open for attention after release.

**Comment by numberZero** [comment]
> discards UDP packets because of flooding

There should be a profiler graph showing lost/resent packets count.

This reminds me a strange behavior I found developing a [CLI Minetest client](https://gitlab.com/numzero/minelight): if you send INIT right after CONNECT, it gets lost (actually, dropped by the server as it isn’t ready yet). A delay is necessary to let the server prepare internal structs; its duration has to be guessed as no ack is sent in reply to CONNECT.

> * In the same function, adding a line to print the timeout each time it's calculated produces an awful lot of "3.000000" even in singleplayer
> * Changing RESEND_TIMEOUT_FACTOR in src/constants.h from 4 (default) to 0.4 fixes singleplayer

*Exactly* 3.0 means that’s nothing but `RESEND_TIMEOUT_MAX`. That means RTT calculated is above 750ms (in singleplayer!). But the fact changing coefficient fixes stuff means it is well beyond 7.5s where nothing would change.

**Comment by SmallJoker** [comment]
#8187 was merged. A real fix is needed, though.
